cli/lsp/testing.rs

```diff
@@ -0,0 +1,1959 @@
+                };
+                if !valid {
+                  if let Some(Ok(parsed_source)) =
+                    document.maybe_parsed_source()
+                  {
+                    let mut collector = TestCollector::new(specifier.clone());
+                    parsed_source.module().visit_with(&mut collector);
+                    let test_definitions = TestDefinitions {
+                      discovered: collector.take(),
+                      injected: Default::default(),
+                      script_version,
+                    };
+                    if !test_definitions.discovered.is_empty() {
+                      client
+                        .send_test_notification(
+                          test_definitions.as_notification(
+                            specifier,
+                            mru.as_ref(),
+                            parsed_source.source(),
+                          ),
+                        )
+                        .await;
+                    }
+                    tests.insert(specifier.clone(), test_definitions);
+                  }
+                }
+              }
+              for key in keys {
+                client
+                  .send_test_notification(as_delete_notification(key))
+                  .await;
+              }
+              performance.measure(mark);
+            }
+          }
+        }
+      })
+    });
+
+    let client = server.client.clone();
+    let runs = server.runs.clone();
+    let mut tokens = HashMap::<u32, CancellationToken>::new();
+    let _run_join_handle = thread::spawn(move || {
+      let runtime = create_basic_runtime();
+
+      runtime.block_on(async {
+        loop {
+          match run_rx.recv().await {
+            None => break,
+            Some(RunRequest::Start(id)) => {
+              let maybe_run = {
+                let runs = runs.lock();
+                runs.get(&id).cloned()
+              };
+              if let Some(run) = maybe_run {
+                let token = {
+                  let token = CancellationToken::new();
+                  tokens.insert(id, token.clone());
+                  token
+                };
+                match run.exec(&client, token, maybe_root_uri.as_ref()).await {
+                  Ok(_) => (),
+                  Err(err) => {
+                    client.show_message(lsp::MessageType::ERROR, err).await;
+                  }
+                }
+                tokens.remove(&id);
+                client
+                  .send_test_notification(TestingNotification::Progress(
+                    lsp_custom::TestRunProgressParams {
+                      id,
+                      message: TestRunProgressMessage::End,
+                    },
+                  ))
+                  .await;
+              }
+            }
+            Some(RunRequest::Cancel(id)) => {
+              if let Some(token) = tokens.get(&id) {
+                token.cancel();
+              }
```

Comment:
I don't think it's possible for the cancellation to occur here. I think we only want to do run `RunRequest::Start` on this thread, so the message to the channel would contain `(id, token)`. Then `TestServer` would hold the tokens and cancel them on the other thread.

Reply:
Yeah, you are right.  I will refactor.

Reply:
I'm not sure why I made it so complicated... I have refactored it and cleaned it up quite a bit.